base/multidimensional.jl

```diff
@@ -94,99 +74,73 @@ immutable CartesianRange{I<:CartesianIndex}
 
-@generated function eachindex(::LinearSlow, A::AbstractArray, B::AbstractArray...)
-    K = max(ndims(A), map(ndims, B)...)
-    startargs = fill(1, K)
-    stopargs = Array{Expr}(K)
-    for i = 1:K
-        Bargs = [:(size(B[$j],$i)) for j = 1:length(B)]
-        stopargs[i] = :(max(size(A,$i),$(Bargs...)))
-    end
-    meta = Expr(:meta, :inline)
-    :($meta; CartesianRange(CartesianIndex{$K}($(startargs...)), CartesianIndex{$K}($(stopargs...))))
-end
+@inline eachindex(::LinearSlow, A::AbstractArray, B::AbstractArray...) = CartesianRange(maxsize((), A, B...))
+maxsize(sz) = sz
+@inline maxsize(sz, A, B...) = maxsize(maxt(sz, size(A)), B...)
+@inline maxt(a::Tuple{}, b::Tuple{}) = ()
+@inline maxt(a::Tuple{}, b::Tuple)   = b
+@inline maxt(a::Tuple,   b::Tuple{}) = a
+@inline maxt(a::Tuple,   b::Tuple)   = (max(a[1], b[1]), maxt(tail(a), tail(b))...)
 
 eltype{I}(::Type{CartesianRange{I}}) = I
 iteratorsize{I}(::Type{CartesianRange{I}}) = Base.HasShape()
 
-@generated function start{I<:CartesianIndex}(iter::CartesianRange{I})
-    N = length(I)
-    cmp = [:(iter.start[$d] > iter.stop[$d]) for d = 1:N]
-    extest = Expr(:||, cmp...)
-    inc = [d < N ? :(iter.start[$d]) : :(iter.stop[$N]+1) for d = 1:N]
-    exstop = :(CartesianIndex{$N}($(inc...)))
-    meta = Expr(:meta, :inline)
-    quote
-        $meta
-        $extest ? $exstop : iter.start
+@inline function start{I<:CartesianIndex}(iter::CartesianRange{I})
+    if any(map(>, iter.start.I, iter.stop.I))
+        return iter.stop+1
     end
+    iter.start
 end
-@generated function next{I<:CartesianIndex}(iter::CartesianRange{I}, state)
-    N = length(I)
-    meta = Expr(:meta, :inline)
-    quote
-        $meta
-        index=state
-        @nif $N d->(index[d] < iter.stop[d]) d->(@nexprs($N, k->(ind_k = ifelse(k>=d, index[k] + (k==d), iter.start[k]))))
-        newindex = @ncall $N $I ind
-        index, newindex
-    end
+@inline function next{I<:CartesianIndex}(iter::CartesianRange{I}, state)
+    state, I(inc(state.I, iter.start.I, iter.stop.I))
 end
-@generated function done{I<:CartesianIndex}(iter::CartesianRange{I}, state)
-    N = length(I)
-    :(state[$N] > iter.stop[$N])
+# increment & carry
+@inline inc(::Tuple{}, ::Tuple{}, ::Tuple{}) = ()
+@inline inc(state::Tuple{Int}, start::Tuple{Int}, stop::Tuple{Int}) = (state[1]+1,)
+@inline function inc(state, start, stop)
+    if state[1] < stop[1]
+        return (state[1]+1,tail(state)...)
+    end
+    newtail = inc(tail(state), tail(start), tail(stop))
+    (start[1], newtail...)
 end
+@inline done{I<:CartesianIndex}(iter::CartesianRange{I}, state) = state.I[end] > iter.stop.I[end]
 
 # 0-d cartesian ranges are special-cased to iterate once and only once
 start{I<:CartesianIndex{0}}(iter::CartesianRange{I}) = false
 next{I<:CartesianIndex{0}}(iter::CartesianRange{I}, state) = iter.start, true
 done{I<:CartesianIndex{0}}(iter::CartesianRange{I}, state) = state
```

Comment:
Would this face the same recent regression as the iteration over numbers which you fixed recently?